### PR TITLE
Support different arguments for different aliases

### LINKED ISSUE #126: Problem with aliases on paginated fields
I'm running into an issue where I have a paginated field with filter arguments that is being [aliased](http://graphql.org/learn/queries/#aliases) to produce two different result trees. The problem is that both result trees end up coming back with the same result (whatever the last alias resolves to).

I can replicate this issue on the [pagination demo](https://join-monster.herokuapp.com/graphql-relay?query=%7B%0A%20%20user(id%3A%202)%20%7B%0A%20%20%20%20first%3A%20posts(first%3A%201)%20%7B%0A%20%20%20%20%20%20edges%20%7B%0A%20%20%20%20%20%20%20%20cursor%0A%20%20%20%20%20%20%20%20node%20%7B%0A%20%20%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20%20%20body%0A%20%20%20%20%20%20%20%20%20%20author%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20fullName%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%20%20second%3A%20posts(first%3A%201%2C%20after%3A%20%22eyJpZCI6NDB9%22)%20%7B%0A%20%20%20%20%20%20edges%20%7B%0A%20%20%20%20%20%20%20%20cursor%0A%20%20%20%20%20%20%20%20node%20%7B%0A%20%20%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20%20%20body%0A%20%20%20%20%20%20%20%20%20%20author%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20fullName%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A) as well.

Here's my query:

```graphql
{
  school {
    privateCourses: resources(first: 1, visibility: private) {
      edges {
        node {
          id
        }
      }
    }
    publicCourses: resources(first: 1, visibility: visible) {
      edges {
        node {
          id
        }
      }
    }
  }
}
```
When this comes back, both `privateCourses` and `publicCourses` are set to the same result (the latter).
```json
{
  "data": {
    "school": {
      "privateCourses": {
        "edges": [
          {
            "node": {
              "id": 18169
            }
          }
        ]
      },
      "publicCourses": {
        "edges": [
          {
            "node": {
              "id": 18169
            }
          }
        ]
      }
    }
  }
}
```

The query join monster generates looks like this:
```sql
SELECT
  "school"."id" AS "id",
  "resources$"."id" AS "resources$__id",
  "resources$"."$total" AS "resources$__$total"
FROM school_school "school"
LEFT JOIN LATERAL (
  SELECT *, count(*) OVER () AS "$total"
  FROM course_course "resources$"
  WHERE "resources$".school_id = "school".id AND "resources$".archived_time is NULL AND "resources$".highest_visibility_level >= 10
  ORDER BY "resources$"."created" DESC
  LIMIT 2 OFFSET 0
) "resources$" ON "resources$".school_id = "school".id
WHERE "school".id = 1
ORDER BY "resources$"."created" DESC
```

Here's the relevant debug output in case it helps:
```
join-monster SQL_AST
  join-monster  { type: 'table',
  name: 'school_school',
  as: 'school',
  children: 
   [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
     { args: { first: 1, visibility: 'visible' },
       paginate: true,
       orderBy: { created: 'desc' },
       type: 'table',
       name: 'course_course',
       as: 'resources$',
       children: 
        [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          { type: 'column',
            name: '$total',
            fieldName: '$total',
            as: '$total' },
          { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          { type: 'column',
            name: '$total',
            fieldName: '$total',
            as: '$total' },
          { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
          { type: 'columnDeps', names: {} } ],
       fieldName: 'resources',
       grabMany: true,
       where: [Function: where],
       sqlJoin: [Function: sqlJoin] },
     { type: 'columnDeps', names: {} } ],
  fieldName: 'school',
  grabMany: false,
  where: [Function: where] } +26ms
  join-monster 
  join-monster SQL
  join-monster  SELECT
  "school"."id" AS "id",
  "resources$"."id" AS "resources$__id",
  "resources$"."$total" AS "resources$__$total"
FROM school_school "school"
LEFT JOIN LATERAL (
  SELECT *, count(*) OVER () AS "$total"
  FROM course_course "resources$"
  WHERE "resources$".school_id = "school".id AND "resources$".archived_time is NULL AND "resources$".highest_visibility_level >= 10
  ORDER BY "resources$"."created" DESC
  LIMIT 2 OFFSET 0
) "resources$" ON "resources$".school_id = "school".id
WHERE "school".id = 1
ORDER BY "resources$"."created" DESC +2ms
```

### LINKED ISSUE #146: Several batch queries of one table
Query:

In this query `onlineFriends` and `offlineFriends` return same data

```graphql
{
  user {
    id
    onlineFriends: friends(online:true) {
      edges {
        node {
          id
        }
      }
    }
    
    offlineFriends: friends(online:false) {
      edges {
        node {
          id
        }
      }
    }
  }
}
```

Schema: 
```javascript
const User = new GraphQLObjectType({
	description: 'A user type',
	name: 'User',
	sqlTable: 'users',
	uniqueKey: 'id',
	fields: () => ({
		id: {
			type: new GraphQLNonNull(GraphQLID)
		},
		
		friends: {
			description: 'A list of user friends',
			sqlPaginate: true,
			type: FriendConnection,
			args: {
				...connectionArgs,
				state: {
					description: 'Current state of friend',
					type: FriendsState
				},
				online: {
					description: 'Online friends field',
					type: GraphQLBoolean
				}
			},
			sortKey: {
				order: 'DESC',
				key: 'date'
			},
			sqlBatch: {
				thisKey: 'user_id',
				parentKey: 'id'
			},
			where(friendsTable, { state, online }) {
				let query = state === 'ALL' ? '' : `${friendsTable}.state = '${state || 'ACCEPTED'}'`
				if (typeof online === 'boolean') {
					query = `${query} AND online = ${online}`
				}

				return query
			}
		}
	})
})

const Friend = new GraphQLObjectType({
	description: 'A friend type',
	name: 'Friend',
	sqlTable: 'friends',
	uniqueKey: ['user_id', 'friend_id'],
	fields: () => ({
		id: {
			type: new GraphQLNonNull(GraphQLID),
			sqlDeps: ['user_id', 'friend_id'],
			resolve: f => `${f.user_id}_${f.friend_id}`
		},
		date: { type: new GraphQLNonNull(GraphQLString) },
		state: { type: new GraphQLNonNull(FriendsState) },
		user: {
			type: new GraphQLNonNull(User),
			sqlJoin: (friendTable, userTable) => `${userTable}.id = ${friendTable}.friend_id`
		},
		online: {
			type: new GraphQLNonNull(GraphQLBoolean)
		}
	})
})
```

Log:

```
  join-monster
  join-monster SQL_AST
  join-monster  { type: 'table',
  name: 'users',
  as: 'user',
  children:
   [ { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
     { type: 'column', name: 'id', fieldName: 'id', as: 'id' },
     { args: { online: false },
       paginate: true,
       sortKey: { order: 'DESC', key: 'date' },
       type: 'table',
       name: 'friends',
       as: 'friends$',
       children:
        [ { type: 'composite',
            name: [ 'user_id', 'friend_id' ],
            fieldName: 'use#fri',
            as: 'use#fri' },
          { type: 'column', name: 'date', fieldName: 'date', as: 'date' },
          { type: 'composite',
            name: [ 'user_id', 'friend_id' ],
            fieldName: 'use#fri',
            as: 'use#fri' },
          { type: 'column', name: 'date', fieldName: 'date', as: 'date' },
          { type: 'columnDeps',
            names: { user_id: 'user_id', friend_id: 'friend_id' } } ],
       fieldName: 'friends',
       grabMany: true,
       where: [Function: where],
       sqlBatch:
        { thisKey:
           { type: 'column',
             name: 'user_id',
             fieldName: 'user_id',
             as: 'user_id' },
          parentKey: { type: 'column', name: 'id', fieldName: 'id', as: 'id' } } },
     { type: 'columnDeps', names: {} } ],
  fieldName: 'user',
  grabMany: false,
  where: [Function: where] } +10s
  join-monster
  join-monster SQL
  join-monster  SELECT
  "user"."id" AS "id"
FROM users "user"
WHERE "user".id = 251586469 +0ms
  join-monster
  join-monster SHAPE_DEFINITION
  join-monster  { id: 'id' } +0ms
SELECT
  "user"."id" AS "id"
FROM users "user"
WHERE "user".id = 251586469
  join-monster
  join-monster RAW DATA
  join-monster  [ anonymous { id: 251586469 } ] +97ms
  join-monster 1 rows... +0ms
  join-monster
  join-monster SHAPED_DATA
  join-monster  { id: 251586469 } +0ms
  join-monster
  join-monster SQL_AST
  join-monster  { args: { online: false },
  paginate: true,
  sortKey: { order: 'DESC', key: 'date' },
  type: 'table',
  name: 'friends',
  as: 'friends$',
  children:
   [ { type: 'composite',
       name: [ 'user_id', 'friend_id' ],
       fieldName: 'use#fri',
       as: 'use#fri' },
     { type: 'column', name: 'date', fieldName: 'date', as: 'date' },
     { type: 'composite',
       name: [ 'user_id', 'friend_id' ],
       fieldName: 'use#fri',
       as: 'use#fri' },
     { type: 'column', name: 'date', fieldName: 'date', as: 'date' },
     { type: 'columnDeps',
       names: { user_id: 'user_id', friend_id: 'friend_id' } },
     { type: 'column',
       name: 'user_id',
       fieldName: 'user_id',
       as: 'user_id' } ],
  fieldName: 'friends',
  grabMany: true,
  where: [Function: where],
  sqlBatch:
   { thisKey:
      { type: 'column',
        name: 'user_id',
        fieldName: 'user_id',
        as: 'user_id' },
     parentKey: { type: 'column', name: 'id', fieldName: 'id', as: 'id' } } } +1ms
  join-monster
  join-monster SQL
  join-monster  SELECT
  NULLIF(CONCAT("friends$"."user_id", "friends$"."friend_id"), '') AS "use#fri",
  "friends$"."date" AS "date",
  "friends$"."user_id" AS "user_id",
  "friends$"."friend_id" AS "friend_id"
FROM (VALUES (251586469)) temp("id")
 JOIN LATERAL (
  SELECT *
  FROM friends "friends$"
  WHERE "friends$"."user_id" = temp."id" AND "friends$".state = 'ACCEPTED' AND online = false
  ORDER BY "friends$"."date" DESC
  LIMIT ALL
) "friends$" ON "friends$"."user_id" = temp."id"
ORDER BY "friends$"."date" DESC +1ms
  join-monster
  join-monster SHAPE_DEFINITION
  join-monster  [ { 'use#fri': 'use#fri',
    date: 'date',
    user_id: 'user_id',
    friend_id: 'friend_id' } ] +0ms
SELECT
  NULLIF(CONCAT("friends$"."user_id", "friends$"."friend_id"), '') AS "use#fri",
  "friends$"."date" AS "date",
  "friends$"."user_id" AS "user_id",
  "friends$"."friend_id" AS "friend_id"
FROM (VALUES (251586469)) temp("id")
 JOIN LATERAL (
  SELECT *
  FROM friends "friends$"
  WHERE "friends$"."user_id" = temp."id" AND "friends$".state = 'ACCEPTED' AND online = false
  ORDER BY "friends$"."date" DESC
  LIMIT ALL
) "friends$" ON "friends$"."user_id" = temp."id"
ORDER BY "friends$"."date" DESC
  join-monster
  join-monster RAW DATA
  join-monster  [ anonymous {
    'use#fri': '251586469121045238',
    date: '2017-06-18T12:38:16.924Z',
    user_id: 251586469,
    friend_id: 121045238 },
  anonymous {
    'use#fri': '251586469166627',
    date: '2015-06-28T20:27:17.000Z',
    user_id: 251586469,
    friend_id: 166627 },
  anonymous {
    'use#fri': '251586469225604',
    date: '2015-06-03T20:58:25.000Z',
    user_id: 251586469,
    friend_id: 225604 },
  anonymous {
    'use#fri': '25158646967315591',
    date: '2014-04-10T21:32:17.000Z',
    user_id: 251586469,
    friend_id: 67315591 } ] +100ms
  join-monster 4 rows... +0ms
  join-monster
  join-monster SHAPED_DATA
  join-monster  [ { 'use#fri': '251586469121045238',
    date: '2017-06-18T12:38:16.924Z',
    user_id: 251586469,
    friend_id: 121045238 },
  { 'use#fri': '251586469166627',
    date: '2015-06-28T20:27:17.000Z',
    user_id: 251586469,
    friend_id: 166627 },
  { 'use#fri': '251586469225604',
    date: '2015-06-03T20:58:25.000Z',
    user_id: 251586469,
    friend_id: 225604 },
  { 'use#fri': '25158646967315591',
    date: '2014-04-10T21:32:17.000Z',
    user_id: 251586469,
    friend_id: 67315591 } ] +1ms
```

### PR DESCRIPTION
**Pull request by kryops**
Hi there,

after experimenting with different workarounds in #481, I believe I may have found a proper fix for #126.

Please let me know what you think :relaxed:

I published the changes to npm as `@kryops/join-monster@3.1.1-1` if anyone wants to try it out with their schema.

### Description

The general strategy is as follows:
* `join-monster` tries to detect conflicting aliases to the same field
* If a conflict is detected, it will set multiple properties on the output object instead of overwriting the `fieldName`:

The query
```graphql
{
  users {
    following { fullName }
    matts: following(name: "matt") { fullName }
  }
}
```

generates per user:

```javascript
{
  following$: [
    { fullName: 'andrew carlson' },
    { fullName: 'matt elder' }
  ],
  following$matts: [
    { fullName: 'matt elder' }
  ],
  following: (args, context, info) => { /* ... */ }
}
```

GraphQL's default resolver supports functions as properties, and will call them with the information necessary to return the correct property.

This is a **breaking change** when using custom resolvers: Instead of always accessing `source[fieldName]`, GraphQL's `defaultFieldResolver` has to be used to get the raw value in order to support multiple conflicting aliases:

```javascript
import { defaultFieldResolver } from 'graphql'

// ...

resolve: (source, args, context, info) => {
  const rawValue = defaultFieldResolver(source, args, context, info)
  return processUsers(rawValue)
}
```

Note that this should be only actually breaking in situations that likely used to return wrong data. For normal fields and non-conflicting aliases the same format is used as before.

### References

Fixes #126

Fixes #146

### Checklist

* [x] I have added documentation for new/changed functionality in this PR via comments and by updating the change log
* [ ] I'm not entirely sure when to consider aliases as "conflicting". For now I have implemented the following logic:
  * Some types of nodes are never conflicting because `join-monster` does not use the args, thus they should point to the same data anyway (e.g. normal columns)
  * Some types of nodes are always conflicting because even using the same args, they can have nested conflicting aliases inside them (e.g. tables, unions)
  * For other types of nodes (e.g. `sqlExpr`), the args may be used by `join-monster`, so aliases are conflicting if they use different args
* [x] Are the generated property keys "unique enough"? What if a schema contains both `comments` and `comments$`?
* [x] We should probably add a separate test for aliases with paginated schemas. Maybe `Post.comments` or `User.comments`?
* [x] We should probably add a separate test for aliases with unions and interfaces. Any ideas how to best do it in the test schema? (should we add args to `writtenMaterial1` or `writtenMaterial2`?)
* [ ] I wasn't sure where to add the new behavior in the documentation. For now, I added it to the warnings page.

### CODE REVIEW

**Comment by DanielSWolf** [comment]
> Are the generated property keys "unique enough"? What if a schema contains both `comments` and `comments$`?

According to the [GraphQL spec](https://spec.graphql.org/June2018/#Name), the `$` character cannot be used in a GraphQL identifier. So it is my understanding that the keys you generate are always unique.

File: test/aliases.js
Comment (DanielSWolf) [inline test/aliases.js]: Did you use `test.only` on purpose here?

Comment (kryops) [reply test/aliases.js]: Oops, good catch! Removed.

**Comment by kryops** [comment]
Just rebased on v3.3.2.

Any chance of moving this forward?

**Comment by nicoabie** [comment]
@bueche can you take a look into this one?

**Comment by nicoabie** [comment]
@kryops This week I become maintainer of this project and I would love helping you merge this PR if you are still interested.

Let me know, have a great weekend!

File: docs/CHANGELOG.md
Comment (bueche) [inline docs/CHANGELOG.md]: assuming this is only required for "non-trivial" field usage. Please confirm. Also, perhaps it's useful to explain in more detail how this helps. Isn't this `defaultFieldResolver()` called under-the-hood?

File: docs/CHANGELOG.md
Comment (bueche) [inline docs/CHANGELOG.md]: and what do you mean by "rawValue" here? raw vs. what??

File: docs/warnings.md
Comment (bueche) [inline docs/warnings.md]: - "non-trivial field" ... this should have a more precise definition at least as best as you understand it.
- "recommend" ? ... perhaps too softly worded. What happens if you don't use the default resolver ... and (again) why are we calling the default resolver _explicitly_ when it was previously called _implicitly_?

File: test-api/schema-basic/User.js
Comment (bueche) [inline test-api/schema-basic/User.js:280]: Excellent to add to the tests ... can you explain what is being added to test this?

File: test/aliases.js
Comment (bueche) [inline test/aliases.js:1]: Really nice set of tests!

**Comment by kryops** [comment]
Hi all, yes I'm still interested in getting this merged.

Thanks for the feedback, I'll try to find some time this week to address it 😅

File: docs/CHANGELOG.md
Comment (kryops) [reply docs/CHANGELOG.md]: Yes, `defaultFieldResolver` is only required for non-trivial fields.

The problem is that GraphQL only uses its `defaultFieldResolver` if you do not specify a custom `resolve` function. As soon as you do, you need to handle this yourself.

I extended the section on custom resolvers in the docs and added a link to it from the changelog, please let me know if this covers it better now 😅

File: docs/CHANGELOG.md
Comment (kryops) [reply docs/CHANGELOG.md]: Good point, it's just the value of the field - changed it to `value`

File: docs/warnings.md
Comment (kryops) [reply docs/warnings.md]: True, it might throw an error or do weird things if you don't use the default resolver and a query uses multiple aliases, because `source[fieldName]` would contain a function instead of the field value.

I changed the wording to "need to", and added a section explaining "non-trivial" values in more detail.

File: test-api/schema-basic/User.js
Comment (kryops) [reply test-api/schema-basic/User.js:280]: This was done to be able to test multiple aliases with different args on union and interface types. I added a comment explaining this.

**Comment by nicoabie** [comment]
I'm planning to do a v3.3.5 with the small fixes currently in master and then create a v3.4.0 release with this.
@bueche what are your thoughts?

**Comment by kryops** [comment]
Should this rather be a v4.0.0 as it is a breaking change for custom resolvers?

**Comment by nicoabie** [comment]
I understood it is only a breaking change for users that are using broken aliases but yeah we can for sure make it v4.0.0

**Comment by bueche** [comment]
hi Nico
will try to look at it soon.
cheers
Ed

File: docs/CHANGELOG.md
Comment (nicoabie) [inline docs/CHANGELOG.md]: can we change "non-trivial" for "fields with multiple aliases"?

File: docs/warnings.md
Comment (nicoabie) [inline docs/warnings.md:108]: on a "multiple-aliased field"

**Review by nicoabie — APPROVED** [review]
PR has been opened for too long. I went over it and it LGTM. It has nice testing suite and code is clean a documented.

I don't want to delay this further, thanks a lot @kryops for you patience and commitment!

File: docs/CHANGELOG.md
Comment (kryops) [reply docs/CHANGELOG.md]: Not sure - whether or not there are multiple aliases depends on the client query, so the schema should probably assume this might always be the case.

What I wanted to express with "non-trivial" is "types of fields that handle multiple aliases this way because they might return different data for each of them" 😅

**Comment by nicoabie** [comment]
Ah ok then link to the definition of non trivial so people reading the
changelog know what it means :)

File: docs/CHANGELOG.md
Comment (nicoabie) [reply docs/CHANGELOG.md]: Ah ok then link to the definition of non trivial so people reading the changelog know what it means 😀

File: docs/warnings.md
Comment (nicoabie) [inline docs/warnings.md]: Link to definition

File: docs/CHANGELOG.md
Comment (kryops) [reply docs/CHANGELOG.md]: Done

File: docs/warnings.md
Comment (kryops) [reply docs/warnings.md]: Done

**Pull request merged**
